ast/compile.go

```diff
@@ -938,10 +938,11 @@ func (c *Compiler) checkRuleConflicts() {
 				childNode, tail := node.find(ref)
 				if childNode != nil {
 					for _, childMod := range childNode.Modules {
-						if childMod.Equal(mod) {
-							continue // don't self-conflict
-						}
 						if len(tail) == 0 {
+							// Avoid recursively checking a module for equality unless we know it's a possible self-match.
+							if childMod.Equal(mod) {
+								continue // don't self-conflict
+							}
```

Comment:
By moving this recursive module equality check behind a guard condition (that can only be true when it's a probable match), we greatly reduce how often the check comes up.

For very large policies (>1500 rules in a package), the savings pile up very rapidly.

Reply:
Rolled the compound-if version into my latest changes.